miden-lib/asm/sat/prologue.masm

```diff
@@ -0,0 +1,363 @@
+
+    # loop over consumed notes and absorb data
+    # ---------------------------------------------------------------------------------------------
+
+    # initialize counter for consumed notes
+    push.0
+
+    # ensure that the number of consumed notes is greater then 0. Conditional for the while loop.
+    dup.1 dup.1 neq
+
+    # loop and absorb note data from the advice provider
+    while.true
+        dup exec.process_consumed_note
+
+        # increment consumed note counter and check if we should loop again
+        push.1 add dup.1 dup.1 neq
+    end
+
+    # drop counter
+    drop
+
+    # compute nullifier commitment
+    # ---------------------------------------------------------------------------------------------
+
+    # initiate counter for nullifiers
+    push.0
+
+    # initiate stack for sequential hash to compute nullifier commitment
+    padw padw padw
+
+    # ensure that the number of consumed notes is greater then 0. Conditional for the while loop.
+    dup.13 dup.13 neq
+
+    # loop and sequentially hash hperm(nullifier, script_root) over all consumed notes
+    while.true
+        # compute address of nullifier
+        dup.12 exec.compute_nullifier_ptr
+
+        # load nullifier from memory and push down stack
+        mem_loadw swapw
+
+        # compute address of script root
+        dup.12 push.1 add push.1024 mul
+        push.CONSUMED_NOTE_DATA_OFFSET.CONSUMED_NOTE_SCRIPT_ROOT_OFFSET add add
+
+        # load script root from memory
+        mem_loadw
+
+        # compute hperm(nullifier, script_root)
+        hperm
+
+        # increment nullifier counter and check if we should loop again
+        movup.12 push.1 add dup movdn.13 dup.14 neq
+    end
+
+    # extract nullifier hash
+    dropw swapw dropw
+
+    # assert nullifier hash is what we would expect
+    padw push.NULLIFIER_COM_PTR mem_loadw assert_eqw
+
+    # clear stack
+    drop drop
+end
+
+# TRANSACTION PROLOGUE
+# =================================================================================================
+
+#! The transaction prologue is executed at the beginning of a transaction. Its responsibility is:
+#! 1. Un-hash public inputs, authenticate the data and store it in the root contexts memory.
+#! 2. Build a single vault containing assets of all inputs (input notes combined with current
+#!    account vault).
+#! 3. Verify that all input notes are present in the note db and that a nullifier does not exist
+#!    for any of them.
+#!
+#! Errors:
+#!  - If data provided by the advice provider does not match public inputs.
+#!  - The account data is invalid.
+#!  - Any of the consumed notes are invalid
+#!
+#! Stack:       [public_inputs]
+#! Advice Tape: [acct_data, consumed_notes_data]
+#! Output:      []
+#!
+#! Note: please see the documentation of the individual procedures for more details.
+proc.transaction_prologue
+    # store public inputs in global inputs section of memory
+    exec.process_public_inputs
```

Comment:
> It seems like the term "public inputs" is somewhat overloaded. I'd probably call these something different as everything except for what goes through the advice provider is a public input. Maybe `env_inputs` or something like that?

Good point, maybe we refer to them as `global_inputs` as specified in the issue?  

> btw, one thing which I think is missing from here is "unhashing" of `pre_block_header` and storing it in memory. We need this so that we have access to `chain_root`, which we can later "unhash" into the peak of the MMR representing the chain.

good catch.  Shall we leave this to a follow up issue? I'm not sure if the structure of the block header is formally defined?

> One question is how the block header hash is computed. We could do it as a simple sequential hash or we could arrange it to make access to `note_root` simpler. In the latter case, block header hash could look like `hash(note_root, sequential_hash_of_everything_else)`.

I think the latter makes sense, making the `note_root` easily accessible will improve efficiency.

Reply:
> Shall we leave this to a follow up issue? I'm not sure if the structure of the block header is formally defined?

Yep - agreed! Maybe leave a `TODO` here - but even that's not required.